#5459 Fix visible 'noname' email attachment file

martgil opened the pull request:
This PR fixes the visible 'noname' attachment file.

Current behavior:
![image](https://github.com/FlowCrypt/flowcrypt-browser/assets/46025304/67575c45-1f2f-4019-a347-45a5eb23fb73)

Proposed fix:
![image](https://github.com/FlowCrypt/flowcrypt-browser/assets/46025304/6b97b857-e3c3-4f0c-b156-ae04ca850f49)

close #5459
close #5460

----------------------------------

**Tests** _(delete all except exactly one)_:
- Tests added or updated

--------------------------------

### To be filled by reviewers

I have reviewed that this PR... _(tick whichever items you personally focused on during this review)_:
- [x] addresses the issue it closes (if any)
- [x] code is readable and understandable
- [x] is accompanied with tests, or tests are not needed
- [x] is free of vulnerabilities
- [x] is documented clearly and usefully, or doesn't need documentation

martgil commented:
@sosnovsky ready for review. thank you!

martgil commented:
I noticed that Live Gmail test is somehow skipped. I'm not sure how I could re-run that particular test.

sosnovsky commented on extension/js/content_scripts/webmail/gmail-element-replacer.ts:
```diff
@@ -457,6 +457,7 @@ export class GmailElementReplacer implements WebmailElementReplacer {
     }
     const loaderContext = new GmailLoaderContext(this.factory, msgEl, attachmentsContainerInner);
     attachmentsContainerInner = $(attachmentsContainerInner);
+    attachmentsContainerInner.parent().find(this.sel.numberOfAttachments).hide();
```
Attachments count label should be visible (as per https://github.com/FlowCrypt/flowcrypt-browser/issues/5460#issuecomment-1784910402)

martgil replied:
Got it - thank you! I understand better why the email `PDF attached` counts 2, because the public key is also part of the count in addition to the pdf attachment.

martgil replied:
We do have the code to show the attachment count label only when the numbers of rendered rendered attachments are more than 2 items. 

I guess this is all good just the count needs to ensure to be correct? It will be fixed then with the later changes.

martgil replied:
@sosnovsky I have fixed the over all issue for `noname` attachment, incorrect number of attachments, and showing the label for attachments. Only the Gmail test, not working quite well.

sosnovsky commented:
> I noticed that Live Gmail test is somehow skipped. I'm not sure how I could re-run that particular test.

We have condition in semaphore.yml to run live tests only in `master` or when branch name ends with `live-test`: https://github.com/FlowCrypt/flowcrypt-browser/blob/ee8fbcc785f0c663f7d738751849d94cb1f57032/.semaphore/semaphore.yml#L62

So live tests don't run in each branch, as then they will interfere between different PRs and fail often.
Another method - comment mentioned line in `semaphore.yml` and live tests will run in this branch.

sosnovsky commented on extension/js/content_scripts/webmail/gmail-element-replacer.ts:
```diff
@@ -473,21 +482,27 @@ export class GmailElementReplacer implements WebmailElementReplacer {
     if (nRenderedAttachments === 0) {
       attachmentsContainerInner.parents(this.sel.attachmentsContainerOuter).first().hide();
+    } else {
+      const elementsToClone = ['span.a2H', 'div.a2b'];
+      const scannedByGmailLabel = $(elementsToClone[0]).first().clone();
+      const scannedByGmailPopover = $(elementsToClone[1]).first().clone(true);
```
maybe we can just update text in existing html elements, instead of cloning them and then re-appending?

martgil replied:
>  I just noticed a small bug 

Thank you - I'll look into it in and will fix that as well.

> maybe we can just update text in existing html elements, instead of cloning them and then re-appending?

Actually, that was my first attempt in fixing the issue but the HTML structuring in Gmail is not scraping-friendly, I mean, the text node and html elements kinda mixed up in a single span so simple text content changing for example via `.text()` don't work very great. It's a bit complex for example, the word/label "One attachment" appears for unmodified number of attachments, wherein with modified number of attachments - they will be defined in a numerical representation which conflicts with the HTML structuring so I came up with element cloning so I can have the better HTML structure I wanted to have + the uniformity of the number of attachments is pleasant compared to modifying "One" to "1" is again complex because the word "One" is a text node and "1" however, is nicely wrapped in an element. 

The live gmail test is not passing at all. Was the test in the web extension page maybe sufficient enough as I could add alternative for testing the expected numbers?

sosnovsky replied:
> Actually, that was my first attempt in fixing the issue but the HTML structuring in Gmail is not scraping-friendly, I mean, the text node and html elements kinda mixed up in a single span so simple text content changing for example via .text() don't work very great. It's a bit complex for example, the word/label "One attachment" appears for unmodified number of attachments, wherein with modified number of attachments - they will be defined in a numerical representation which conflicts with the HTML structuring so I came up with element cloning so I can have the better HTML structure I wanted to have + the uniformity of the number of attachments is pleasant compared to modifying "One" to "1" is again complex because the word "One" is a text node and "1" however, is nicely wrapped in an element.

Got it, thanks for detailed explanation, then let's use your current implementation 👍

> The live gmail test is not passing at all. Was the test in the web extension page maybe sufficient enough as I could add alternative for testing the expected numbers?

Having live gmail test will be useful for detecting if there will be any changes in Google's html markup, which can lead to breaking of our html parsing. We can use existing tests, instead of creating a separate one:
- `mail.google.com - attachments which contain emoji in filename are rendered correctly` - here let's check that page has label `2 Attachments  •  Scanned by Gmail`
- `mail.google.com - pubkey file gets rendered` - here check that attachments count label is hidden

martgil replied:
> Works much better now, well done, I just noticed a small bug

Thank you @sosnovsky. It is now fixed.

 I have made changes that would satisfy all this. Test are all passing now.

https://flowcrypt.semaphoreci.com/workflows/5ebecf93-c789-40ad-888d-ce64a720642e?pipeline_id=2fa93122-d051-44d3-855d-94cd5319d44b

sosnovsky reviewed (commented):
Works much better now, well done, I just noticed a small bug - please check `flowcrypt-browser #5410 attached ecc public key` thread in `flowcrypt.compatibility` inbox, it doesn't hide attachments count label:

<img width="400" alt="Screenshot 2023-11-06 at 14 12 27" src="https://github.com/FlowCrypt/flowcrypt-browser/assets/1062093/0e123076-850e-4b19-84ab-99281ddfa859">

martgil commented:
@sosnovsky ready for review. thank you!

sosnovsky reviewed (commented):
I checked `flowcrypt-browser #5410 attached ecc public key` thread and attachments count label is hidden now, but `import public key` block is hidden too - it should be visible, as we should hide only attachments count label in this case.

Also please check `Encrypted email with public key attached` thread in `flowcrypt.compatibility` account, it has the same issue as reported in previous review - `import public key` and attachments count label are both visible:

<img width="600" alt="Screenshot 2023-11-09 at 11 19 40" src="https://github.com/FlowCrypt/flowcrypt-browser/assets/1062093/7f835e5b-552c-4241-a87f-8442f93b1172">


Both mentioned threads should show `import public key` block and hide attachments count label.

martgil commented:
> Also please check Encrypted email with public key attached thread in flowcrypt.compatibility account

It working well for me - no label shown. Only the import public key button is missing.

![image](https://github.com/FlowCrypt/flowcrypt-browser/assets/46025304/fb2ed48e-c901-435b-9ce2-1a4285309a15)

sosnovsky commented:
> It working well for me - no label shown.

Seems like it's a different thread - on my screenshot sender is `e2e.enterprise.test`(thread id is `FMfcgzGtwzspbBBstksRxMDLgCHFjHhS`), while on your screenshot - `flowcrypt.compatibility`:

<img width="780" alt="Screenshot 2023-11-10 at 13 22 11" src="https://github.com/FlowCrypt/flowcrypt-browser/assets/1062093/3fe2efe4-61c4-4396-b407-e4d8c7020719">

sosnovsky commented:
> I'll be checking it once again. it is actually an encrypted public key, so it was once detected as regular encrypted attachment then later been recognized and rendered as public key from which the process of hiding labels comes first. So i would need to find a way to recognized it as encrypted public key before hand so when I reached the test nRenderedattachments, it will let the rendering of the public key ui.

It works well now, I just noticed that top separator isn't hidden on `public key` block:

<img width="500" alt="Screenshot 2023-11-10 at 13 25 12" src="https://github.com/FlowCrypt/flowcrypt-browser/assets/1062093/767e30af-cac2-4274-bb8d-f7a13bd30571">

martgil commented:
@sosnovsky, I have also addressed the last reported issues. Could you please take another look?

sosnovsky reviewed (commented):
> It works well now, I just noticed that top separator isn't hidden on public key block:

This issue is fixed ✅

> Seems like it's a different thread - on my screenshot sender is e2e.enterprise.test(thread id is FMfcgzGtwzspbBBstksRxMDLgCHFjHhS), while on your screenshot - flowcrypt.compatibility:

This issue is still presented - `1 Attachment  •  Scanned by Gmail` label isn't hidden in `FMfcgzGtwzspbBBstksRxMDLgCHFjHhS` thread (checked in Chrome and Firefox)

martgil commented:
> This issue is still presented

I have fixed it now and also fixed this case in pgp_block.

Before:
![image](https://github.com/FlowCrypt/flowcrypt-browser/assets/46025304/f1edfa26-c9fb-468a-bb26-f62e2e2da65d)

After:
![image](https://github.com/FlowCrypt/flowcrypt-browser/assets/46025304/1358333d-e79f-46ea-a64e-d54631a88368)

sosnovsky reviewed (approved):
All good now, thanks for fixes!

martgil commented:
> All good now, thanks for fixes!

Thank you too for all the help.
